File: modules/sessioncontrol.ts
//
//---------------- Session Control Window -------------
//
import { Widget, Utils, PopupWindow, Roundedges, Gdk } from "imports"
import icons from "lib/icons.js"
import options from "options"
import { winheight } from "lib/screensizeadjust"

const { RoundedAngleEnd } = Roundedges
const { Button, Box, Label, Revealer, Icon, EventBox } = Widget
const { pwrmenu } = options

type Action = "lock" | "reboot" | "logout" | "shutdown"

class PowerMenu extends Service {
	static {
		Service.register(this, {}, {
			"cmd": ["string"],
		})
	}
	#cmd = ""
	get cmd() { return this.#cmd }
	action(action: Action) {
		[this.#cmd] = {
			lock: [pwrmenu.lock.value],
			reboot: [pwrmenu.reboot.value],
			logout: [pwrmenu.logout.value],
			shutdown: [pwrmenu.shutdown.value],
		}[action]
		App.closeWindow("sessioncontrols")
	}
	readonly shutdown = () => {
		this.action("shutdown")
	}
}

const powermenu = new PowerMenu

const SysButton = (action: Action, label: string) => Button({
	on_clicked: () => powermenu.action(action) || Utils.execAsync(powermenu.cmd),
	child: Box({
		vertical: true,
		vpack: "center",
		hpack: "center",
		className: "sessioncontrol-btn",
		children: [
			Icon(icons.powermenu[action]),
			Label({
				label,
				visible: pwrmenu.labels.bind(),
			}),
		],
	}),
})

export default () => PopupWindow({
	name: "sessioncontrols",
	className: "sessioncontrols",
	anchor: ["top", "left", "right", "bottom"],
	transition: "slide_up",
	child: Box({
		className: "sessioncontrols-firstbox",
		hpack: "center",
		vpack: "end",
		vexpand: true,
		css: `margin-top: ${winheight(0.765)}px;`,
		setup: self => self.hook(pwrmenu.layout, () => {
			self.toggleClassName("box", pwrmenu.layout.value === "box")
			self.toggleClassName("line", pwrmenu.layout.value === "line")
		}),
		children: [
			RoundedAngleEnd("bottomleft", {class_name: "angleLarge"}),
			
			Box({
				vpack: "end",
				className: "sessioncontrols-box",
				hpack: "center",
				hexpand: false,
				vexpand: false,
				spacing: 30,
				children:[
					SysButton("lock", "Lock"),
					SysButton("logout", "Log Out"),
					SysButton("reboot", "Reboot"),
					SysButton("shutdown", "Shutdown"),
				]
			}),

			RoundedAngleEnd("bottomright", {class_name: "angleLarge"}),
		]
	}),
})

File: Rust/Modules/Windows/SessionControl.rs
//
//---------------- Session Control Window -------------
//
use imports::{Widget, Utils, PopupWindow, Roundedges, Gdk, Service, App};
use lib::icons;
use options;
use lib::screensizeadjust::winheight;

use std::collections::HashMap;

type Action = String;

struct PowerMenu {
    cmd: String,
}

impl PowerMenu {
    fn new() -> Self {
        PowerMenu {
            cmd: String::new(),
        }
    }

    fn action(&self, action: &str) {
        let pwrmenu = options::pwrmenu();
        let cmd = match action {
            "lock" => pwrmenu.lock.value.clone(),
            "reboot" => pwrmenu.reboot.value.clone(),
            "logout" => pwrmenu.logout.value.clone(),
            "shutdown" => pwrmenu.shutdown.value.clone(),
            _ => String::new(),
        };
        App::close_window("sessioncontrols");
        Utils::exec_async(&cmd);
    }

    fn shutdown(&self) {
        self.action("shutdown");
    }
}

static POWER_MENU: PowerMenu = PowerMenu::new();

fn sys_button(action: &str, label: &str) -> Widget {
    let power_menu = &POWER_MENU;
    Widget::Button(Box::new(move |button| {
        button.on_clicked(Box::new(move || {
            power_menu.action(action);
        }));
        Box::new({
            let mut children = Vec::new();
            children.push(Widget::Icon(icons::powermenu(action)));
            children.push(Widget::Label {
                label: label.to_string(),
                visible: options::pwrmenu().labels.bind(),
            });
            children
        })
    })
}

fn session_controls() -> PopupWindow {
    let pwrmenu = options::pwrmenu();
    PopupWindow {
        name: "sessioncontrols".to_string(),
        class_name: "sessioncontrols".to_string(),
        anchor: ["top".to_string(), "left".to_string(), "right".to_string(), "bottom".to_string()],
        transition: "slide_up".to_string(),
        child: Box::new({
            let mut children = Vec::new();
            children.push(Roundedges::RoundedAngleEnd("bottomleft".to_string(), HashMap::new()));
            children.push(Box::new({
                let mut children = Vec::new();
                children.push(Box::new({
                    let mut children = Vec::new();
                    children.push(sys_button("lock", "Lock"));
                    children.push(sys_button("logout", "Log Out"));
                    children.push(sys_button("reboot", "Reboot"));
                    children.push(sys_button("shutdown", "Shutdown"));
                    children
                }));
                children
            }));
            children.push(Roundedges::RoundedAngleEnd("bottomright".to_string(), HashMap::new()));
            children
        }),
    }
}